### ISSUE TITLE
Proposal: Catch::Tags(...)

### ISSUE THREAD

**Issue by mwpowellhtx** [philsquared/Catch#1062]
Hello,

I wrote a set of *Catch::Tags(...)* variadic template functions to help support a thing I wanted to do: namely, present my tags as individually named, or as vectors of tags, and so on, without necessarily concerning myself with the *[square][bracket][syntax]*. Thus far it's working out great.

Thought y'all might like to take a look. Currently it is in [header](http://github.com/mwpowellhtx/cppnngswig/blob/master/swig/cpp/tests/catch/catch_tags.h) and [source](http://github.com/mwpowellhtx/cppnngswig/blob/master/swig/cpp/tests/catch/catch_tags.cpp) form, but perhaps it can exist as header only, I'm not sure.

And with a [usage example](http://github.com/mwpowellhtx/cppnngswig/blob/master/swig/cpp/tests/transport/transport.cpp#L124).

Cheers.

**Comment by horenmar** [comment]
Got here from 1095.

FYI your implementation invokes UB, as the constructed `Tags` are out of scope after the test case is constructed, but test case doesn't take ownership of tags (as the most common expected usage is with static strings) . This means that any attempt to access the tag ref from test case afterwards is invalid.

**Comment by mwpowellhtx** [comment]
@horenmar I'm not sure what you mean by "in scope" after the test case is constructed. I'm not attempting to reference any tags after test case construction, nor is that the intended purpose. I am using ``Catch::Tags`` as a helper to *[build][the][syntax]* and that's it. Afterwards, Catch runs with the tag string as-is, how ever they are provided. Given that as the stated objective, I'm not sure how that can be considered "UB".

One caveat, however, is that *Catch v1* uses **const char*** for test case parameters, so the caller must invoke *std::string().c_str()*.  In *Catch v2*, that's promoted to a *std::string* friendly construct, the *c_str()* requirement goes away.

**Comment by horenmar** [comment]
I investigated it further, and I was wrong.

Both parameters in `TEST_CASE` macro are provided to `NameAndTags` object, which takes `StringRef`s\* as arguments to the constructor. `StringRef` does not take ownership of the string, meaning that if the originating string goes away, it keeps pointing to invalid memory. This means that once a `TEST_CASE` is created, the `NameAndDesc` object points to invalid memory, when given `Catch::Tags`.

Except that the implementation internals are rather old, crufty and the first thing they do with a created `NameAndDesc`, is to (rather pointlessly) convert its members into separate `std::string`s. Because this happens during the `TEST_CASE` construction, it catches the `StringRef`s while they are still valid, instead of later on.

If the implementation was optimized better, it would keep things as a `StringRef` until it is necessary not to\** and the contents would be accessed once invalid.


\* For Catch 1, it was `const char*`, but the principle is the same.
\** Depending on what you optimize for, you might even prefer to keep them as `StringRef`s forever and either keep paying the reparsing cost, or have the tags be `StringRef` into a substring of the full tag string.

**Comment by mwpowellhtx** [comment]
@horenmar Yes, good points. For a use case such as these, ownership of the memory is tricky. I'd consider "test case" the owner of the memory, once it is passed to whatever receives the buffer. If there's another behavior, I don't know what it is. All I know is it seemed to work just fine, at least for *v1*. Untested in *v2*, but I don't see why it wouldn't work, assuming consistent ownership principles. My two cents.
